clang/lib/Sema/SemaChecking.cpp

```diff
@@ -11427,6 +11427,19 @@ static void AnalyzeImplicitConversions(
     return;
   }
 
+  if (auto *OutArgE = dyn_cast<HLSLOutArgExpr>(E)) {
+    // The base expression is only used to initialize the parameter for
+    // arguments to `inout` parameters, so we only traverse down the base
+    // expression for `inout` cases.
```

Comment:
I should have commented on this. If SemaChecking traverses down both sides we get warnings for the conversion that would "initialize" the `out` parameter, but those warnings aren't accurate because no initialization occurs.

If we only traverse down the write back side we only get the warning for the copy back to the l-value, which is the behavior we want.

clang/test/SemaHLSL/Language/OutputParameters.hlsl verifies the behavior for `out` arguments of only diagnosing the write back conversion, removing the condition breaks that test by emitting diagnostics on the initialization as well.

Reply:
I specifically said the original l-value expression.  I'm trying to make sure we get warnings in sub-expressions of the l-value that have nothing to do with the final conversion to produce the argument.

With that said:
> If SemaChecking traverses down both sides we get warnings for the conversion that would "initialize" the out parameter, but those warnings aren't accurate because no initialization occurs.

Isn't this an argument that you either (1) shouldn't visit the initialization specifically for `out`, but should still do it for `inout`, or (2) shouldn't generate the initialization expression for `out`?

Apropos, I know this is a future direction, but what semantics do you want for C++ class types and `out`?  Skipping the initialization entirely is definitely incorrect — the callee will get passed a reference to an uninitialized object and (presumably) start calling methods on it.  Probably this should default-initialize?

Reply:
Gotcha! I think I misunderstood your comment since I didn't have the original lvalue preserved initially. I'll update this.

These are pretty wonky when we start looking at C++isms. My team reacted with disgust when I posed this potential (but currently invalid) code:

```hlsl
RWBuffer<int> Val;

struct A {
  int Idx;
  A &operator ++() {
    Val[Idx] = Val[Idx] + 1;
    return *this;
  }
};

void init(out A Obj) {
  Obj.Idx = 2;
};

[numthreads(1,1,1)]
void main() {
  A a;
  a.Idx = 0;
  init(++a);
  ++a;
}
```

We certainly need to evaluate the argument expression (so that it can side-effect). Constructors would make this more complicated, but it probably should default initialize. Are you okay with that being a follow-up? If so I'll file an issue to track doing that work.

Reply:
Yeah, since you're not supporting C++ types here, it's fine to leave the work for later.